File: concepts/07_modules_packages_and_loading/03_initialization_caching_cycles_and_dynamic_loading/r/test_02_failed_import_and_cache_state.R
# polyglot-family: modules_packages_and_loading
# polyglot-concept: initialization_caching_cycles_and_dynamic_loading
# polyglot-related: languages/r/tooling_and_runtime/14_packages_namespaces_and_tools/
# polyglot-related+: test_106_namespace_library_require_and_colon_operators.R
#
# 共同问题：加载失败是否留下半初始化 cache，后续探测怎样表现。
# 对照观察：`requireNamespace(..., quietly=TRUE)` 对不存在 package 返回 FALSE，不加入 loaded namespace。

name <- "polyglot.package.that.does.not.exist"
before <- loadedNamespaces()
available <- requireNamespace(name, quietly = TRUE)
after <- loadedNamespaces()

stopifnot(
    identical(available, FALSE),
    !(name %in% after),
    identical(before, after)
)

File: concepts/07_modules_packages_and_loading/03_initialization_caching_cycles_and_dynamic_loading/lua/test_02_failed_import_and_cache_state.lua
-- Common question: what cache state remains after failed or cyclic initialization?
-- Inputs: a failing preload loader, retry, and two explicitly cycle-safe modules.
-- Observations: failure is not cached, retries rerun, manual early cache entry, and shared cycle identity.
-- polyglot-family: modules_packages_and_loading
-- polyglot-concept: initialization_caching_cycles_and_dynamic_loading
-- polyglot-related: languages/lua/tooling_and_runtime/13_modules_and_loading/
-- polyglot-related+: test_099_require_searchers_cache_failures_and_cycles.lua

local t = require("support.assertions")
local attempts = 0
package.preload.failed_module = function()
    attempts = attempts + 1
    error("load failed")
end
package.loaded.failed_module = nil

t.raises(function() require("failed_module") end, "load failed")
t.equal(package.loaded.failed_module, nil)
t.raises(function() require("failed_module") end, "load failed")
t.equal(attempts, 2)

package.preload.cycle_a = function()
    local exports = {name = "a"}
    package.loaded.cycle_a = exports
    exports.other = require("cycle_b")
    return exports
end
package.preload.cycle_b = function()
    return {name = "b", other = require("cycle_a")}
end
local cycle = require("cycle_a")
t.equal(cycle.other.name, "b")
t.same(cycle.other.other, cycle)

t.done()
